### PR TITLE
Support ALTER TABLE REBUILD PARTITION syntax

### PR DESCRIPTION
**Pull request by spongedu**
### What problem does this PR solve? 
for #448 , support `ALTER TABLE REBUILD PARTITION` syntax in `Parser`

### What is changed and how it works?
1. Add related parser rule in parser.y
2. Complete Restore method for struct AlterTableSpec
3. Add related tests

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 N / A

Side effects

 N / A

### CODE REVIEW

**Comment by kennytm** [comment]
Is this blocked by #453 (same situation as #457)?

**Comment by spongedu** [comment]
> Is this blocked by #453 (same situation as #457)?

@kennytm yep, seems so. All these issues require some structure like `AllOrPartitionList`

**Comment by spongedu** [comment]
@kennytm PTAL

File: parser.y [hunk 1]
```diff
@@ -4004,7 +4022,7 @@ UnReservedKeyword:
  "ACTION" | "ASCII" | "AUTO_INCREMENT" | "AFTER" | "ALWAYS" | "AVG" | "BEGIN" | "BIT" | "BOOL" | "BOOLEAN" | "BTREE" | "BYTE" | "CLEANUP" | "CHARSET" %prec charsetKwd
 | "COLUMNS" | "COMMIT" | "COMPACT" | "COMPRESSED" | "CONSISTENT" | "CURRENT" | "DATA" | "DATE" %prec lowerThanStringLitToken| "DATETIME" | "DAY" | "DEALLOCATE" | "DO" | "DUPLICATE"
 | "DYNAMIC" | "ENCRYPTION" | "END" | "ENFORCED" | "ENGINE" | "ENGINES" | "ENUM" | "ERRORS" | "ESCAPE" | "EXECUTE" | "FIELDS" | "FIRST" | "FIXED" | "FLUSH" | "FOLLOWING" | "FORMAT" | "FULL" |"GLOBAL"
-| "HASH" | "HOUR" | "INSERT_METHOD" | "LESS" | "LOCAL" | "LAST" | "NAMES" | "OFFSET" | "PASSWORD" %prec lowerThanEq | "PREPARE" | "QUICK" | "REDUNDANT"
+| "HASH" | "HOUR" | "INSERT_METHOD" | "LESS" | "LOCAL" %prec local | "LAST" | "NAMES" | "OFFSET" | "PASSWORD" %prec lowerThanEq | "PREPARE" | "QUICK" | "REBUILD" | "REDUNDANT"
```
Comment (leoppro) [inline parser.y]:
```suggestion
| "HASH" | "HOUR" | "INSERT_METHOD" | "LESS" | "LOCAL" | "LAST" | "NAMES" | "OFFSET" | "PASSWORD" %prec lowerThanEq | "PREPARE" | "QUICK" | "REBUILD" | "REDUNDANT"
```

File: parser.y [hunk 2]
```diff
@@ -1404,6 +1405,23 @@ AlterTableSpec:
 		yylex.AppendError(yylex.Errorf("The DISCARD PARTITION TABLESPACE clause is parsed but ignored by all storage engines."))
 		parser.lastErrorAsWarn()
 	}
+|	"REBUILD" "PARTITION" NoWriteToBinLogAliasOpt AllOrPartitionNameList
+	{
+		yylex.AppendError(yylex.Errorf("REBUILD PARTITION syntax is parsed but not implement for now."))
```
Comment (leoppro) [inline parser.y]: remove L1408-1410

Comment (spongedu) [reply parser.y]: ok

**Comment by spongedu** [comment]
@leoppro  PTAL

**Comment by leoppro** [comment]
resolve the conflicts please

**Comment by spongedu** [comment]
> resolve the conflicts please

Done

**Pull request merged**